Experimental CMake support

g-easy opened the pull request:
This is addressing issue #86 

Rolling my changes up as a PR to make reviews easier.

g-easy commented:
Two comments from @coryan that weren't resolved on the commit:

1. Namespacing: a later commit namespaced all the OpenCensus targets. Is this sufficient for now?

g-easy commented:
2. I get googletest via ExternalProject, and only if OpenCensus is being built with tests (i.e. when developing OC, not when developing an app that uses OC). Is this reasonable for now?

coryan commented on cmake/README.md:
````diff
@@ -0,0 +1,27 @@
+```shell
+cmake -H. -B.build
+cmake --build .build
+cmake --build .build --target test
````
That only works with `make` as the generator, with things like `MSBuild` it does not.  This works with all generators:

```
(cd .build && ctest --output-on-failure)
```

Well, on Windows you need to add `-C Debug` or `-C Release` for "reasons".

g-easy replied:
Thanks, I've updated the docs! I'm still pretty new to CMake.

coryan commented on examples/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# add_subdirectory(grpc) TODO
```
Instead of using `grpc` as a subdirectory consider using it via `externalproject_add` like you do for Abseil.  That way you do not get their targets in your namespace, **and** you would be testing a configuration that more closely resembles what you get via `find_package()`.

coryan commented on cmake/OpenCensusDeps.cmake:
```diff
@@ -0,0 +1,60 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+if(BUILD_TESTING)
```
Why not use `externalproject_add()` for googletest too?

g-easy replied:
I thought I was. If BUILD_TESTING is enabled, I do the usual dance starting on L19 with `configure_file`, which uses `cmake/googletest.CMakeLists.txt` which has the externalproject_add() call.

meastp commented:
Hi,

This is great, thank you for working on this! :) 

I have experimented with adding opencensus-cpp to vcpkg and building it. Currently only static opencensus-cpp configurations are supported.

I then use vcpkg's version of gtest and abseil, both of which are included with find_package.

I have added:

- ```cmake
  IF(MSVC)
    add_definitions(-DNOMINMAX) 
  ENDIF()
  ```
  This fixes compilation errors on MSVC.
- Replaced the `${OPENCENSUS_INCLUDE_DIR}` with the built in `${CMAKE_SOURCE_DIR}` (not `${CMAKE_CURRENT_SOURCE_DIR}`)
- I have added support for find_package for opencensus-cpp ( `install( TARGETS ... )` and `install ( EXPORT ...)`.
  When doing this I also added include files for the targets ( added a HEADERS argument when using the `opencensus_lib` function).

I would like to:
- Add google benchmarks (vcpkg also supports google benchmark, so it should be easy to add).
- Precisely specify which link libraries are PRIVATE to the implementation and PUBLIC (exposed to clients) when using target_link_libraries.
- With support for find_package added, I think the add_library(${PROJECT_NAME}::${NAME} ALIAS ${_NAME}) should be removed.

How do you feel about this? Which changes are you willing to merge? :)

coryan commented:
@meastp 

> * ```cmake
>   IF(MSVC)
>     add_definitions(-DNOMINMAX) 
>   ENDIF()
>   ```

Shouldn't that be `target_compile_definitions()` for the right targets?

> * I have added support for find_package for opencensus-cpp ( `install( TARGETS ... )` and `install ( EXPORT ...)`.

:+1:

meastp commented:
> 
> 
> @meastp
> 
> > * ```cmake
> >   IF(MSVC)
> >     add_definitions(-DNOMINMAX) 
> >   ENDIF()
> >   ```
> 
> Shouldn't that be `target_compile_definitions()` for the right targets?
> 

Ah, you're right, of course, thanks :) I'm reading up on modern cmake to be able to contribute here, and I overlooked this function.

I think that the distinction between PRIVATE and PUBLIC library link dependencies from each target is needed for this to make sense, though. (I can do this if there is interest, but I want to know which changes are acceptable to you first - see previous reply/message)

g-easy commented:
> I have experimented with adding opencensus-cpp to vcpkg and building it. Currently only static opencensus-cpp configurations are supported.
>
> I then use vcpkg's version of gtest and abseil, both of which are included with find_package.

Awesome!

> I have added support for find_package for opencensus-cpp ( install( TARGETS ... ) and install ( EXPORT ...).
> When doing this I also added include files for the targets ( added a HEADERS argument when using the opencensus_lib function).

This is great, opencensus-cpp needs to support `find_package` at some point.

> I would like to:
> 
> Add google benchmarks (vcpkg also supports google benchmark, so it should be easy to add).

I'd definitely like Benchmark support (I think there's a TODO for it). Is there an ExternalProject version of this for people who don't use vcpkg?

> Precisely specify which link libraries are PRIVATE to the implementation and PUBLIC (exposed to clients) when using target_link_libraries.

I think part of this is done, at least with respect to which libraries we expose in the ::namespace, right?

> With support for find_package added, I think the add_library(${PROJECT_NAME}::${NAME} ALIAS ${_NAME}) should be removed.

I don't know enough about CMake yet to be able to make an informed decision here, let's discuss it.

> How do you feel about this? Which changes are you willing to merge? :)

This sounds great, and I'm willing to consider all of it. I'm going to have some questions about some of the changes.

If I merge this PR soon, would you be willing to send PRs for the stuff above?

meastp commented:
> 
> This sounds great, and I'm willing to consider all of it. I'm going to have some questions about some of the changes.
> 
> If I merge this PR soon, would you be willing to send PRs for the stuff above?

Great, I'll send PRs for the changes when this is merged, and we can discuss each change separately (and in more detail) :)

meastp commented:
I'm wondering if using [object libraries](https://cmake.org/cmake/help/latest/command/add_library.html#object-libraries) for some of the targets/components would be more appopriate? There are quite a lot of libraries (.lib) generated on build. What do you think?

coryan commented:
@meastp @g-easy should we move this discussion to a separate bug?

> I'm wondering if using object libraries for some of the targets/components would be more appopriate? There are quite a lot of libraries (.lib) generated on build. What do you think?

If you folks decide to use object libraries remember that position-indepedent code is not enabled by default for them:

https://cmake.org/cmake/help/v3.13/prop_tgt/POSITION_INDEPENDENT_CODE.html

You have to explicitly set the property if you are building them into shared objects.  My recommendation is that you make the first round of changes you proposed, and do the object library changes (if @g-easy approves) afterwards.

meastp commented:
> 
> 
> @meastp @g-easy should we move this discussion to a separate bug?
> 
> > I'm wondering if using object libraries for some of the targets/components would be more appopriate? There are quite a lot of libraries (.lib) generated on build. What do you think?
> 
> If you folks decide to use object libraries remember that position-indepedent code is not enabled by default for them:
> 
> https://cmake.org/cmake/help/v3.13/prop_tgt/POSITION_INDEPENDENT_CODE.html
> 
> You have to explicitly set the property if you are building them into shared objects. My recommendation is that you make the first round of changes you proposed, and do the object library changes (if @g-easy approves) afterwards.

Sure, I'm just wondering if there is another way to structure the targets that do not result in so many library files (especially for shared library builds). I'm not opposed to delaying that discussion 'till after the rest of the changes.

g-easy commented:
@meastp I've opened #244 to track all the followups suggested in this PR. :)
